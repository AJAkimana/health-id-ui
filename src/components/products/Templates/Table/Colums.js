/* eslint-disable import/prefer-default-export */
export const columns = [
  {
    name: 'id',
    label: 'Id',
    options: {
      filter: false,
      sort: true,
      display: false,
    }
  },
  {
    name: 'productName',
    label: 'Name',
    options: {
      filter: true,
      sort: true,
      display: true,
    }
  },
  {
    name: 'skuNumber',
    label: 'SKU',
    options: {
      filter: true,
      sort: true,
      display: true,
    }
  },
  {
    name: 'category',
    label: 'Category',
    options: {
      filter: true,
      sort: true,
      display: true,
    }
  },
  {
    name: 'measurementUnit',
    label: 'Dispensing_Size',
    options: {
      filter: true,
      sort: true,
      display: true,
    }
  },
  {
    name: 'brand',
    label: 'Brand',
    options: {
      filter: true,
      sort: true,
      display: true,
    }
  },
  {
    name: 'manufacturer',
    label: 'Manufacturer',
    options: {
      filter: true,
      sort: true,
      display: true,
    }
  },
  {
    name: 'productQuantity',
    label: 'QT_In_Stock',
    options: {
      filter: true,
      sort: true,
      display: false,
    }
  },
  {
    name: 'salesPrice',
    label: 'Sales_Price',
    options: {
      filter: true,
      sort: true,
      display: true,
    }
  },
  {
    name: 'reorderPoint',
    label: 'Reorder_Point',
    options: {
      filter: true,
      sort: true,
      display: true,
    }
  },
  {
    name: 'reorderMax',
    label: 'Reorder_Max',
    options: {
      filter: true,
      sort: true,
      display: true,
    }
  },
  {
    name: 'loyaltyWeight',
    label: 'Loyalty_Weight',
    options: {
      filter: true,
      sort: true,
      display: true,
    }
  },
  {
    name: 'preferredSupplier',
    label: 'Preferred_Supplier',
    options: {
      filter: true,
      sort: true,
      display: true,
    }
  },
  {
    name: 'backupSupplier',
    label: 'Backup_Supplier',
    options: {
      filter: true,
      sort: true,
      display: true,
    }
  },
  {
    name: 'nearestExpiryDate',
    label: 'Nearest_Expiry_Date',
    options: {
      filter: true,
      sort: true,
      display: true,
    }
  },
  {
    name: 'vatStatus',
    label: 'VAT_Status',
    options: {
      filter: true,
      sort: true,
      display: true,
    }
  },
];
