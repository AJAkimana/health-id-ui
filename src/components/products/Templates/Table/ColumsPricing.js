/* eslint-disable import/prefer-default-export */
export const columns = [{
  name: 'id',
  label: 'Id',
  options: {
    filter: false,
    sort: true,
    display: false,
  }
},
{
  name: 'productName',
  label: 'Name',
  options: {
    filter: true,
    sort: true,
    display: true,
  }
},
{
  name: 'skuNumber',
  label: 'SKU',
  options: {
    filter: true,
    sort: true,
    display: true,
  }
},
{
  name: 'salesPrice',
  label: 'Sale_Price',
  options: {
    filter: true,
    sort: true,
    display: true,
  }
},
{
  name: 'markup',
  label: 'MarkUp',
  options: {
    filter: true,
    sort: true,
    display: true,
  }
},
{
  name: 'batchInfo',
  label: 'Latest_Batch_Unit_Cost',
  options: {
    filter: true,
    sort: true,
    display: true,
  }
},
{
  name: 'GrossMargin',
  label: 'Gross_Margin',
  options: {
    filter: true,
    sort: true,
    display: true,
  }
},
{
  name: 'category',
  label: 'Category',
  options: {
    filter: true,
    sort: true,
    display: true,
  }
},
{
  name: 'loyaltyWeight',
  label: 'Loyalty_WGT',
  options: {
    filter: true,
    sort: true,
    display: true,
  }
},
{
  name: 'vatStatus',
  label: 'VAT_Status',
  options: {
    filter: true,
    sort: true,
    display: true,
  }
},
];
