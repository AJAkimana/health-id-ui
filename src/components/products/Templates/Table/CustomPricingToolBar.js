import React, { Component, Fragment } from 'react';
import { withStyles } from '@material-ui/core/styles';
import PropTypes from 'prop-types';
import {
  Tooltip, IconButton
} from '@material-ui/core';
import exportlogo from '../../../../assets/images/products/export.png';
import addlogo from '../../../../assets/images/products/add.png';
import searchlogo from '../../../../assets/images/products/search.png';
import { ToolbarStyles } from '../../../../assets/styles/stock/stock';
import TableSearch from '../../../stock_control/Table/TableSearch';
import { CustomIconButton } from '../../../stock_control/utils/utils';

export class CustomPricingToolBar extends Component {
  state = {
    open: false,
    addSupplierOpen: false
  };

  handleToggle = () => {
    const { open } = this.state;
    this.setState({ open: !open });
  };

  handleToggleViewMenu = () => {
    const { openViewMenu } = this.state;
    this.setState({ openViewMenu: !openViewMenu });
  };

  handleClose = () => {
    this.setState({ openAddMenu: false, openViewMenu: false });
  };

  render() {
    const { openViewMenu } = this.state;
    const {
      classes,
      handleClickSearch,
      isSearchActive,
      handleHideSearch,
      handleSearchTextChange,
    } = this.props;
    return (
      <Fragment>
        {isSearchActive ? (
          <TableSearch onHide={handleHideSearch} handleTextChange={handleSearchTextChange} />
        ) : ('')}
        <Fragment>
          <Tooltip title="Search" style={{ marginRight: '25px' }}>
            <IconButton
              className={!isSearchActive ? classes.iconButtonActive : classes.iconButton}
              buttonRef={(node) => {
                this.anchorEl = node;
              }}
              toolTip="Search"
              aria-owns={openViewMenu ? 'menu-list-grow' : undefined}
              aria-haspopup="true"
              onClick={handleClickSearch}
            >
              <img src={searchlogo} style={{ width: '20px' }} alt="" />
            </IconButton>
          </Tooltip>
          <CustomIconButton
            toolTip="Add Product"
            buttonRef={(node) => {
              this.addProductElement = node;
            }}

          >
            <img src={addlogo} style={{ width: '20px' }} alt="" />
          </CustomIconButton>
          <Tooltip title="Export List" style={{ marginRight: '25px' }}>
            <IconButton>
              <img src={exportlogo} style={{ width: '20px' }} alt="" />
            </IconButton>
          </Tooltip>
        </Fragment>
      </Fragment>
    );
  }
}

CustomPricingToolBar.propTypes = {
  classes: PropTypes.instanceOf(Object).isRequired,
  handleClickSearch: PropTypes.func.isRequired,
  handleHideSearch: PropTypes.func.isRequired,
  isSearchActive: PropTypes.bool,
  handleSearchTextChange: PropTypes.func.isRequired,
};

CustomPricingToolBar.defaultProps = {
  isSearchActive: false
};

export default withStyles(ToolbarStyles)(CustomPricingToolBar);
