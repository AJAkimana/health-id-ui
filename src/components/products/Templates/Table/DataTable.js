/* eslint-disable react/prefer-stateless-function */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import {
  Checkbox,
  Paper,
  TableRow,
  TablePagination,
  TableCell,
  TableBody,
  Table
} from '@material-ui/core';
import TableToolBar from './TableToolBar';
import { TableStyles } from '../../../../assets/styles/stock/stock';
import DataTableLoader from '../../../dataTable/dataTableLoader';
import TableHeader from './TableHeader';
import { ProductInfoPopup } from '../ProductInfoPopup';
import { getProductInformationCardPosition } from '../../../utils/screen';
import { getSortedData } from '../../filter';

export class DataTable extends Component {
  state = {
    selected: [],
    hoverdItem: undefined,
    x: 0,
    y: 0,
    isSearchActive: false,
    data: [],
    rows: [],
    isSearching: false,
    order: 'asc',
    orderBy: 'name',
  }

  static getDerivedStateFromProps(props, state) {
    if (state.isSearching) {
      return null;
    }
    return { data: props.data, rows: props.data };
  }

  isSelected = (name) => {
    const { selected } = this.state;
    return selected.indexOf(name) !== -1;
  };

  handleRowSeleted = (_, id) => {
    const { selected } = this.state;
    const selectedIndex = selected.indexOf(id);
    let newSelected = [];

    if (selectedIndex === -1) {
      newSelected = newSelected.concat(selected, id);
    } else if (selectedIndex === 0) {
      newSelected = newSelected.concat(selected.slice(1));
    } else if (selectedIndex === selected.length - 1) {
      newSelected = newSelected.concat(selected.slice(0, -1));
    } else if (selectedIndex > 0) {
      newSelected = newSelected.concat(
        selected.slice(0, selectedIndex),
        selected.slice(selectedIndex + 1)
      );
    }
    this.setState({ selected: newSelected });
  };

  handleHideSearch = () => {
    this.setState({ isSearchActive: false });
  }

  handleOnRowHover = (e, row) => {
    const { x, y } = getProductInformationCardPosition(e);
    this.setState({ hoverdItem: row, x, y });
  }

  handleHidePopup = () => {
    this.setState({ hoverdItem: undefined });
  }

  handleClickSearch = () => {
    const { isSearchActive } = this.state;
    this.setState({ isSearchActive: !isSearchActive });
  }

  handleRequestSort = (_, property) => {
    const { order, orderBy } = this.state;
    const isDesc = orderBy === property && order === 'desc';
    this.setState({ order: isDesc ? 'asc' : 'desc', orderBy: property });
  };

  handleSelectAllClick = (event) => {
    const { rows } = this.state;
    if (event.target.checked) {
      const newSelections = rows.map(product => product.id);
      this.setState({ selected: newSelections });
      return;
    }
    this.setState({ selected: [] });
  };


  render() {
    const {
      classes,
      columns,
      title,
      onRowClick,
      isAdmin,
      status,
      handleViewProposed,
      handleChangeRowsPerPage,
      totalCount,
      handleChangePage,
      pageNumber,
      rowsCount,
      handleSearch,
      loading,
      client,
    } = this.props;
    const {
      selected,
      hoverdItem,
      x,
      y,
      isSearchActive,
      rows,
      order,
      orderBy,
    } = this.state;
    return (
      <div>
        <Paper className={classes.paper}>
          <TableToolBar
            name="toolbar"
            isAdmin={isAdmin}
            title={`${title}`}
            numSelected={selected.length}
            client={client}
            handleSearchTextChange={e => handleSearch(e, client)}
            isSearchActive={isSearchActive}
            handleHideSearch={this.handleHideSearch}
            handleClickSearch={this.handleClickSearch}
            status={status}
            handleViewProposed={handleViewProposed}
          />
          {
            loading ? (<DataTableLoader />)
              : (
                <div className={classes.tableWrapper}>
                  <Table
                    onMouseLeave={this.handleHidePopup}
                    className={classes.table}
                    aria-labelledby="tableTitle"
                  >
                    <TableHeader
                      numSelected={selected.length}
                      order={order}
                      orderBy={orderBy}
                      onSelectAllClick={this.handleSelectAllClick}
                      onRequestSort={this.handleRequestSort}
                      rowCount={rows.length}
                      headRows={columns}
                    />
                    {
                      hoverdItem && (
                        <ProductInfoPopup
                          onForwardButtonClick={onRowClick}
                          handleHidePopup={this.handleHidePopup}
                          position={{ x, y }}
                          row={hoverdItem}
                          classes={classes}
                        />
                      )
                    }
                    <TableBody>
                      {rows.length > 0
                        ? getSortedData(rows, order, orderBy)
                          .map((row) => {
                            const isItemSelected = this.isSelected(row.id);
                            return (
                              <TableRow
                                role="checkbox"
                                aria-checked={isItemSelected}
                                tabIndex={-1}
                                onMouseEnter={e => this.handleOnRowHover(e, row)}
                                key={row.id}
                                style={TableStyles.tableRow}
                                selected={isItemSelected}
                                onClick={() => {
                                  onRowClick(row.id);
                                }}
                              >
                                <TableCell padding="checkbox">
                                  <Checkbox
                                    checked={isItemSelected}
                                    onClick={(event) => {
                                      event.stopPropagation();
                                      this.handleRowSeleted(event, row.id);
                                    }}
                                  />
                                </TableCell>
                                {
                                  columns.map(col => (col.options.display !== false && (<TableCell align="left">{row[col.name]}</TableCell>)))
                                }

                              </TableRow>
                            );
                          }) : (<TableRow><TableCell align="center" colSpan={14}>No Products</TableCell></TableRow>)
                      }
                    </TableBody>
                  </Table>
                </div>

              )
          }

          <TablePagination
            rowsPerPageOptions={[5, 10, 25]}
            component="div"
            count={totalCount}
            rowsPerPage={rowsCount}
            page={pageNumber - 1}
            backIconButtonProps={{
              'aria-label': 'Previous Page'
            }}
            nextIconButtonProps={{
              'aria-label': 'Next Page'
            }}
            onChangePage={handleChangePage}
            onChangeRowsPerPage={handleChangeRowsPerPage}
          />
        </Paper>
      </div>
    );
  }
}

DataTable.propTypes = {
  classes: PropTypes.instanceOf(Object).isRequired,
  columns: PropTypes.arrayOf(String).isRequired,
  title: PropTypes.string.isRequired,
  isAdmin: PropTypes.bool.isRequired,
  onRowClick: PropTypes.func.isRequired,
  handleViewProposed: PropTypes.func.isRequired,
  status: PropTypes.string.isRequired,
  totalCount: PropTypes.number.isRequired,
  pageNumber: PropTypes.number.isRequired,
  rowsCount: PropTypes.number.isRequired,
  handleChangePage: PropTypes.func.isRequired,
  handleChangeRowsPerPage: PropTypes.func.isRequired,
  handleSearch: PropTypes.func.isRequired,
  loading: PropTypes.bool.isRequired,
  client: PropTypes.instanceOf(Object).isRequired
};

export default withStyles(TableStyles)(DataTable);
