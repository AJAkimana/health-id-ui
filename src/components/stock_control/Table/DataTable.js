import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import {
  Fade, Popper, Checkbox, Paper, TableRow,
  TablePagination, TableCell, TableBody, Table
} from '@material-ui/core';

import ProductCard from '../ProductCard';
import TableHeader from './TableHeader';
import TableToolBar from './TableToolBar';
import { TableStyles } from '../../../assets/styles/stock/stock';
import {
  stableSort, getSorting
} from '../utils/utils';

export const DataTable = ({ classes, columns, data, title, onRowClick }) => {
  const [anchorEl, setAnchorEl] = useState(null);
  const [isSearchActive, setIsSearchActive] = useState(false);
  const [prevSearchValue, setPrevSearchValue] = useState('');
  const [openPopper, setOpenPopper] = useState(false);
  const [order, setOrder] = useState('asc');
  const [rows, setRows] = useState(data);
  const [hoveredItem, setHoveredItem] = useState({});
  const [orderBy, setOrderBy] = useState('name');
  const [selected, setSelected] = useState([]);
  const [page, setPage] = useState(0);
  const [rowsPerPage, setRowsPerPage] = useState(10);

  const handleRequestSort = (_, property) => {
    const isDesc = orderBy === property && order === 'desc';

    setOrder(isDesc ? 'asc' : 'desc');
    setOrderBy(property);
  };

  const handleSelectAllClick = (event) => {
    if (event.target.checked) {
      const newSelections = rows.map(product => product.id);
      setSelected(newSelections);
      return;
    }
    setSelected([]);
  };

  const handleClick = (_, id) => {
    const selectedIndex = selected.indexOf(id);
    let newSelected = [];

    if (selectedIndex === -1) {
      newSelected = newSelected.concat(selected, id);
    } else if (selectedIndex === 0) {
      newSelected = newSelected.concat(selected.slice(1));
    } else if (selectedIndex === selected.length - 1) {
      newSelected = newSelected.concat(selected.slice(0, -1));
    } else if (selectedIndex > 0) {
      newSelected = newSelected.concat(
        selected.slice(0, selectedIndex),
        selected.slice(selectedIndex + 1)
      );
    }

    setSelected(newSelected);
  };

  const handleChangePage = (_, newPage) => {
    setPage(newPage);
  };

  const handleChangeRowsPerPage = (event) => {
    setRowsPerPage(+event.target.value);
  };

  const handleTextChange = (event) => {
    let searchResults = [];
    const searchText = event.target.value;
    const searchMatch = new RegExp(searchText, 'i');

    if (searchText !== prevSearchValue && searchText.length > 0) {
      setPrevSearchValue(searchText);
      searchResults = data.filter(
        row => searchMatch.test(row.name)
        || searchMatch.test(row.sku)
      );
      setRows(searchResults);
    } else {
      setRows(data);
      setPrevSearchValue('');
    }
  };

  const handlePopperClick = (event) => {
    const hoverData = rows.find(value => value.name === event.currentTarget.innerText);

    setAnchorEl(event.currentTarget);
    setHoveredItem(hoverData);
    setOpenPopper(true);
  };

  const handleClickInverseSelection = () => {
    const newSelected = [];

    rows.forEach((row) => {
      if (!selected.find(selectedId => selectedId === row.id)) {
        newSelected.push(row.id);
      }
    });
    setSelected(newSelected);
  };

  const isSelected = name => selected.indexOf(name) !== -1;

  const emptyRows = rowsPerPage - Math.min(rowsPerPage, rows.length - page * rowsPerPage);

  return (
    <div className={classes.root}>
      <Paper className={classes.paper}>
        <TableToolBar
          title={`${rows.length} ${title}`}
          numSelected={selected.length}
          handleTextChange={handleTextChange}
          handleClickDeselectAll={() => {
            setSelected([]);
          }}
          isSearchActive={isSearchActive}
          handleHideSearch={() => setIsSearchActive(false)}
          handleClickSearch={() => setIsSearchActive(!isSearchActive)}
          handleClickInverseSelection={() => handleClickInverseSelection()}
        />
        <div className={classes.tableWrapper}>
          <Table className={classes.table} aria-labelledby="tableTitle">
            <TableHeader
              numSelected={selected.length}
              order={order}
              orderBy={orderBy}
              onSelectAllClick={handleSelectAllClick}
              onRequestSort={handleRequestSort}
              rowCount={rows.length}
              headRows={columns}
            />
            <TableBody>
              {stableSort(rows, getSorting(order, orderBy))
                .slice(page * rowsPerPage, page * rowsPerPage + rowsPerPage)
                .map((row) => {
                  const isItemSelected = isSelected(row.id);
                  const {
                    id, name, sku, measurementunit, quantity
                  } = row;
                  return (
                    <TableRow
                      role="checkbox"
                      aria-checked={isItemSelected}
                      tabIndex={-1}
                      key={id}
                      selected={isItemSelected}
                      onClick={() => {
                        onRowClick(id);
                      }}
                    >
                      <TableCell padding="checkbox">
                        <Checkbox
                          checked={isItemSelected}
                          onClick={(event) => {
                            event.stopPropagation();
                            handleClick(event, id);
                          }}
                        />
                      </TableCell>
                      <TableCell align="left">
                        <span
                          onMouseEnter={(event) => {
                            handlePopperClick(event);
                          }}
                          onMouseLeave={() => {
                            setOpenPopper(!openPopper);
                          }}
                          className={classes.name}
                        >
                          {name}
                        </span>
                      </TableCell>
                      <TableCell align="left">{sku}</TableCell>
                      <TableCell align="left">{measurementunit}</TableCell>
                      <TableCell align="left">{quantity}</TableCell>
                    </TableRow>
                  );
                })}
              {emptyRows > 0 && (
                <TableRow style={{ height: 49 * emptyRows }}>
                  <TableCell colSpan={6} />
                </TableRow>
              )}
            </TableBody>
          </Table>
        </div>
        <TablePagination
          rowsPerPageOptions={[5, 10, 25]}
          component="div"
          count={rows.length}
          rowsPerPage={rowsPerPage}
          page={page}
          backIconButtonProps={{
            'aria-label': 'Previous Page'
          }}
          nextIconButtonProps={{
            'aria-label': 'Next Page'
          }}
          onChangePage={handleChangePage}
          onChangeRowsPerPage={handleChangeRowsPerPage}
        />
      </Paper>
      <Popper
        placement="right-start"
        transition
        disablePortal
        open={openPopper}
        anchorEl={anchorEl}
      >
        {({ TransitionProps }) => (
          <Fade {...TransitionProps} timeout={350}>
            <Paper>
              <ProductCard data={hoveredItem} />
            </Paper>
          </Fade>
        )}
      </Popper>
    </div>
  );
};

DataTable.propTypes = {
  classes: PropTypes.instanceOf(Object).isRequired,
  columns: PropTypes.arrayOf(String).isRequired,
  data: PropTypes.arrayOf(Object).isRequired,
  title: PropTypes.string.isRequired,
  onRowClick: PropTypes.func.isRequired
};

export default withStyles(TableStyles)(DataTable);
