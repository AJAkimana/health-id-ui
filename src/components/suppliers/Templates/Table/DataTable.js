import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import {
  Checkbox,
  Paper,
  TableRow,
  TablePagination,
  TableCell,
  TableBody,
  Table
} from '@material-ui/core';
import NoteIcon from '@material-ui/icons/Note';
import { supplyStyles } from '../../../../assets/styles/suppliers/suppliers';
import Rating from '../Rating';
import TableHeader from '../../../stock_control/Table/TableHeader';
import TableToolBar from './TableToolBar';
import { TableStyles } from '../../../../assets/styles/stock/stock';
import { stableSort, getSorting } from '../../../stock_control/utils/utils';

export const DataTable = ({
  classes, columns, data, title, onRowClick, isAdmin
}) => {
  const [setAnchorEl] = useState(null);
  const [isSearchActive, setIsSearchActive] = useState(false);
  const [prevSearchValue, setPrevSearchValue] = useState('');
  const [openPopper, setOpenPopper] = useState(false);
  const [order, setOrder] = useState('asc');
  const [rows, setRows] = useState(data);
  const [setHoveredItem] = useState({});
  const [orderBy, setOrderBy] = useState('name');
  const [selected, setSelected] = useState([]);
  const [page, setPage] = useState(0);
  const [rowsPerPage, setRowsPerPage] = useState(10);

  const handleRequestSort = (_, property) => {
    const isDesc = orderBy === property && order === 'desc';

    setOrder(isDesc ? 'asc' : 'desc');
    setOrderBy(property);
  };

  useEffect(() => {
    setRows(data);
  });

  const handleSelectAllClick = (event) => {
    if (event.target.checked) {
      const newSelections = rows.map(product => product.id);
      setSelected(newSelections);
      return;
    }
    setSelected([]);
  };

  const handleClick = (_, id) => {
    const selectedIndex = selected.indexOf(id);
    let newSelected = [];

    if (selectedIndex === -1) {
      newSelected = newSelected.concat(selected, id);
    } else if (selectedIndex === 0) {
      newSelected = newSelected.concat(selected.slice(1));
    } else if (selectedIndex === selected.length - 1) {
      newSelected = newSelected.concat(selected.slice(0, -1));
    } else if (selectedIndex > 0) {
      newSelected = newSelected.concat(
        selected.slice(0, selectedIndex),
        selected.slice(selectedIndex + 1)
      );
    }

    setSelected(newSelected);
  };

  const handleChangePage = (_, newPage) => {
    setPage(newPage);
  };

  const handleChangeRowsPerPage = (event) => {
    setRowsPerPage(+event.target.value);
  };

  const handleTextChange = (event) => {
    let searchResults = [];
    const searchText = event.target.value;
    const searchMatch = new RegExp(searchText, 'i');

    if (searchText !== prevSearchValue && searchText.length > 0) {
      setPrevSearchValue(searchText);
      searchResults = data.filter(row => searchMatch.test(row.name) || searchMatch.test(row.sku));
      setRows(searchResults);
    } else {
      setRows(data);
      setPrevSearchValue('');
    }
  };

  const handlePopperClick = (event) => {
    const hoverData = rows.find(value => value.name === event.currentTarget.innerText);

    setAnchorEl(event.currentTarget);
    setHoveredItem(hoverData);
    setOpenPopper(true);
  };

  const handleClickInverseSelection = () => {
    const newSelected = [];

    rows.forEach((row) => {
      if (!selected.find(selectedId => selectedId === row.id)) {
        newSelected.push(row.id);
      }
    });
    setSelected(newSelected);
  };

  const isSelected = name => selected.indexOf(name) !== -1;

  const handleNotes = (notes) => {
    const notesCount = notes.length;
    if (notesCount >= 1) {
      return (<NoteIcon style={supplyStyles.noteStyle} />);
    }
  };

  const emptyRows = rowsPerPage - Math.min(rowsPerPage, rows.length - page * rowsPerPage);

  return (
    <div className={classes.root}>
      <Paper className={classes.paper}>
        <TableToolBar
          name="toolbar"
          isAdmin={isAdmin}
          title={`${rows.length} ${title}`}
          numSelected={selected.length}
          handleTextChange={handleTextChange}
          handleClickDeselectAll={() => {
            setSelected([]);
          }}
          isSearchActive={isSearchActive}
          handleHideSearch={() => {
            setIsSearchActive(false);
            setRows(data);
            setPrevSearchValue('');
          }}
          handleClickSearch={() => setIsSearchActive(!isSearchActive)}
          handleClickInverseSelection={() => handleClickInverseSelection()}
        />

        <div className={classes.tableWrapper}>
          <Table className={classes.table} aria-labelledby="tableTitle">
            <TableHeader
              numSelected={selected.length}
              order={order}
              orderBy={orderBy}
              onSelectAllClick={handleSelectAllClick}
              onRequestSort={handleRequestSort}
              rowCount={rows.length}
              headRows={columns}
            />
            <TableBody>
              {stableSort(rows, getSorting(order, orderBy))
                .slice(page * rowsPerPage, page * rowsPerPage + rowsPerPage)
                .map((row) => {
                  const isItemSelected = isSelected(row.id);
                  const {
                    id, name, rating, tier, notes
                  } = row;
                  return (
                    <TableRow
                      role="checkbox"
                      aria-checked={isItemSelected}
                      tabIndex={-1}
                      key={id}
                      selected={isItemSelected}
                      onClick={() => {
                        onRowClick(id);
                      }}
                    >
                      <TableCell padding="checkbox">
                        <Checkbox
                          checked={isItemSelected}
                          onClick={(event) => {
                            event.stopPropagation();
                            handleClick(event, id);
                          }}
                        />
                      </TableCell>
                      <TableCell align="left">{id}</TableCell>
                      <TableCell align="left">
                        <span
                          onMouseEnter={(event) => {
                            event.stopPropagation();
                            handlePopperClick(event);
                          }}
                          onMouseLeave={(event) => {
                            event.stopPropagation();
                            setOpenPopper(!openPopper);
                          }}
                          className={classes.name}
                        >
                          {name}
                        </span>
                      </TableCell>
                      <TableCell align="left">{tier}</TableCell>
                      <TableCell align="left"><Rating rating={rating} /></TableCell>
                      <TableCell align="left">
                        {handleNotes(notes)}
                      </TableCell>
                    </TableRow>
                  );
                })}
              {emptyRows > 0 && (
                <TableRow style={{ height: 49 * emptyRows }}>
                  <TableCell colSpan={6} />
                </TableRow>
              )}
            </TableBody>
          </Table>
        </div>

        {/* Pagination for the table */}

        <TablePagination
          rowsPerPageOptions={[5, 10, 25]}
          component="div"
          count={rows.length}
          rowsPerPage={rowsPerPage}
          page={page}
          backIconButtonProps={{
            'aria-label': 'Previous Page'
          }}
          nextIconButtonProps={{
            'aria-label': 'Next Page'
          }}
          onChangePage={handleChangePage}
          onChangeRowsPerPage={handleChangeRowsPerPage}
        />
      </Paper>

      {/* Use popper here, and also view detail */}
    </div>
  );
};

DataTable.propTypes = {
  classes: PropTypes.instanceOf(Object).isRequired,
  columns: PropTypes.arrayOf(String).isRequired,
  data: PropTypes.arrayOf(Object).isRequired,
  title: PropTypes.string.isRequired,
  isAdmin: PropTypes.bool.isRequired,
  onRowClick: PropTypes.func.isRequired
};

export default withStyles(TableStyles)(DataTable);
