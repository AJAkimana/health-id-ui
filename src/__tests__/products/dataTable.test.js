import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { DataTable } from '../../components/products/Templates/Table/DataTable';
describe('Supplier Page DataTable ', () => {
  const event = {
    stopPropagation: jest.fn(),
    currentTarget: {
      innerText: 1
    },
    target: {
      value: 'panadol'
    }
  };
  const props = {
    data: [],
    classes: {},
    columns: [],
    title: 'rfrfr',
    onRowClick: () => {},
    isAdmin: false,
    status: 'approved',
    handleViewProposed: () => {},
    totalCount: () => {},
    handleChangePage: () => {},
    pageNumber: 3,
    rowsCount: 10,
    handleSearch: () => {},
    loading: false,
    client: {
      query: () => {}
    },
  }
  const props2 = {
    data: [],
    classes: {},
    columns: [],
    title: 'rfrfr',
    onRowClick: () => {},
    isAdmin: false,
    status: 'approved',
    handleViewProposed: () => {},
    totalCount: () => {},
    handleChangePage: () => {},
    pageNumber: 3,
    rowsCount: 10,
    handleSearch: () => {},
    loading: true,
    client: {
      query: () => {}
    },
  }
  const props3 = {
    data: [
      {
        "id": "261",
        "productName": "Panadol",
        "skuNumber": "000261",
        "description": "Nice meds, they mess you real good",
        "brand": "Stans",
        "manufacturer": "Stans",
        "vatStatus": false,
        "productQuantity": 85,
        "salesPrice": 408.0,
        "nearestExpiryDate": "2019-08-13",
        "loyaltyWeight": 5,
        "tags": [],
        "productCategory": {
          "id": "15",
          "name": "pain killer"
        },
        "measurementUnit": {
          "id": "1",
          "name": "tablets"
        },
        "preferredSupplier": {
          "id": "2",
          "name": "sean2"
        },
        "backupSupplier": {
          "id": "2",
          "name": "sean2"
        }
      }
    ],
    classes: {},
    columns: [],
    title: 'rfrfr',
    onRowClick: () => {},
    isAdmin: false,
    status: 'approved',
    handleViewProposed: () => {},
    totalCount: () => {},
    handleChangePage: () => {},
    pageNumber: 3,
    rowsCount: 10,
    handleSearch: () => {},
    loading: true,
    client: {
      query: () => {}
    },
  }

  it('renders without crashing', () => {
    const wrapper = shallow(<DataTable {...props} />);
    const wrapper2 = shallow(<DataTable {...props2} />);
    const wrapper3= shallow(<DataTable {...props3} />);
    wrapper3.instance().setState({ isSearching: true, selected: ['xnl', 'hop'] });
    wrapper3.instance().isSelected('xnl');
    wrapper3.instance().handleRowSeleted(null, 'xnl');
    wrapper3.instance().setState({ isSearching: true, selected: { indexOf: () => -1 } });
    wrapper3.instance().handleRowSeleted(null, 'xnl');
    wrapper3.instance().setState({ isSearching: true, selected: { indexOf: () => 1, length: 1, slice: () => {} } });
    wrapper3.instance().handleRowSeleted(null, 'xnl');
    wrapper3.instance().setState({ isSearching: true, selected: { indexOf: () => 1, length: 2, slice: () => {} } });
    wrapper3.instance().handleRowSeleted(null, 'xnl');
    wrapper2.instance().handleOnRowHover({ pageX: 345, pageY: 343 }, props3.data);
    wrapper2.instance().handleOnRowHover({ pageX: 900, pageY: 843 }, props3.data);
    wrapper.instance().handleHidePopup();
    wrapper.instance().handleClickSearch();
    wrapper3.instance().handleHideSearch();
    wrapper3.instance().setState({ order: 'asc', orderBy: 'name' });
    wrapper.instance().handleRequestSort(null, 'name');
    wrapper3.instance().setState({ order: 'desc', orderBy: 'name' });
    wrapper.instance().handleRequestSort(null, 'name');
    wrapper3.instance().handleSelectAllClick({ target: { checked: true } });
    wrapper3.instance().handleSelectAllClick({ target: { checked: false } });
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
