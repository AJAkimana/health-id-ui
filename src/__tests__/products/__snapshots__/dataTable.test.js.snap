// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Supplier Page DataTable  renders without crashing 1`] = `
<div>
  <WithStyles(Paper)>
    <TableToolBar
      client={
        Object {
          "query": [Function],
        }
      }
      handleClickSearch={[Function]}
      handleEdit={[Function]}
      handleHideSearch={[Function]}
      handleSearchTextChange={[Function]}
      handleViewProposed={[Function]}
      isAdmin={false}
      isSearchActive={true}
      name="toolbar"
      numSelected={0}
      status="approved"
      title="rfrfr"
    />
    <div>
      <WithStyles(Table)
        aria-labelledby="tableTitle"
        onMouseLeave={[Function]}
      >
        <WithStyles(TableHeader)
          headRows={Array []}
          numSelected={0}
          onRequestSort={[Function]}
          onSelectAllClick={[Function]}
          order="asc"
          orderBy="name"
          rowCount={0}
        />
        <WithStyles(TableBody)>
          <WithStyles(TableRow)>
            <WithStyles(TableCell)
              align="center"
              colSpan={14}
            >
              No Products
            </WithStyles(TableCell)>
          </WithStyles(TableRow)>
        </WithStyles(TableBody)>
      </WithStyles(Table)>
    </div>
    <WithStyles(TablePagination)
      backIconButtonProps={
        Object {
          "aria-label": "Previous Page",
        }
      }
      component="div"
      count={[Function]}
      nextIconButtonProps={
        Object {
          "aria-label": "Next Page",
        }
      }
      onChangePage={[Function]}
      page={2}
      rowsPerPage={10}
      rowsPerPageOptions={
        Array [
          5,
          10,
          25,
        ]
      }
    />
  </WithStyles(Paper)>
</div>
`;
